ddtrace/contrib/asyncio/provider.py

```diff
@@ -1,74 +1,89 @@
 import asyncio
 
-from ...provider import BaseContextProvider
-from ...provider import DatadogContextMixin
+from ...compat import PYTHON_VERSION_INFO
+from ...provider import DefaultContextProvider
 from ...span import Span
+from .compat import asyncio_current_task
 
 
-class AsyncioContextProvider(BaseContextProvider, DatadogContextMixin):
-    """Manages the active context for asyncio execution. Framework
-    instrumentation that is built on top of the ``asyncio`` library, should
-    use this provider when contextvars are not available (Python versions
-    less than 3.7).
+class BaseAsyncioContextProvider(DefaultContextProvider):
+    def _executor_id(self):
+        # type: () -> int
+        return id(asyncio_current_task())
 
-    This Context Provider inherits from ``DefaultContextProvider`` because
-    it uses a thread-local storage when the ``Context`` is propagated to
-    a different thread, than the one that is running the async loop.
-    """
 
-    # Task attribute used to set/get the context
-    _CONTEXT_ATTR = "__datadog_context"
+if PYTHON_VERSION_INFO >= (3, 7, 0):
 
-    def activate(self, context, loop=None):
-        """Sets the scoped ``Context`` for the current running ``Task``."""
-        loop = self._get_loop(loop)
-        if not loop:
-            super(AsyncioContextProvider, self).activate(context)
+    class AsyncioContextProvider(BaseAsyncioContextProvider):
+        pass
+
+
+else:
+
+    class AsyncioContextProvider(BaseAsyncioContextProvider):
+        """Manages the active context for asyncio execution. Framework
+        instrumentation that is built on top of the ``asyncio`` library, should
+        use this provider when contextvars are not available (Python versions
+        less than 3.7).
+
+        This Context Provider inherits from ``DefaultContextProvider`` because
+        it uses a thread-local storage when the ``Context`` is propagated to
+        a different thread, than the one that is running the async loop.
+        """
+
+        # Task attribute used to set/get the context
+        _CONTEXT_ATTR = "__datadog_context"
+
+        def activate(self, context, loop=None):
+            """Sets the scoped ``Context`` for the current running ``Task``."""
+            loop = self._get_loop(loop)
+            if not loop:
+                super(AsyncioContextProvider, self).activate(context)
+                return context
+
+            # the current unit of work (if tasks are used)
+            task = asyncio.Task.current_task(loop=loop)
+            if task:
+                setattr(task, self._CONTEXT_ATTR, context)
```

Comment:
do we need the `isinstance(context, Span)` check here to make a copy?

Reply:
This should be handled by ` _update_active` below in `active()` which handles it in the general case